src/net.cpp

```diff
@@ -2597,7 +2599,18 @@ uint64_t CConnman::GetTotalBytesSent()
 
 ServiceFlags CConnman::GetLocalServices() const
 {
-    return nLocalServices;
+    uint64_t local_services = nLocalServices;
+    if (local_services & NODE_COMPACT_FILTERS) {
```

Comment:
The point of this code is that the filter index may be building in the background and we don't want to signal NODE_COMPACT_FILTERS to peers until it is in sync. Then as soon as it does finish syncing, turn it on. 

I'm not sure this is the best way to do this so, I'm open to suggestions (this is the subject of two of the questions in the PR description). Another option would be to have a separate CLI flag for signalling NODE_COMPACT_FILTERS, which also blocks the net threads until the filter indices are in sync. That way the user can sync the filter index in the background, then shut down the node and restart with the BIP 157 net stuff enabled. Thoughts?

Reply:
I spent a few minutes writing an indignant comment about how we should be okay with taking this small hack on as technical debt, but it actually looks like it might be pretty easy to address @TheBlueMatt's concerns by adding a method to `CConman` (eg `AddLocalService(ServiceFlags flag)`) that just modifies `nLocalServices`. The indexing code could call it on sync completion. Any reason not to do that?

Reply:
(FWIW I hate the idea of blocking net threads while waiting for filter indices to sync, and slightly less hate requiring a restart to serve the filters.)

Reply:
@jamesob Yes, that would work too, but I wouldn't want to index to call CConmann directly. It should do it through some sort of callback to avoid coupling them, which just creates more indirection. I think this approach is a bit easier, but I'd be OK with a callback that updates `nLocalServices` as well.

Reply:
Just had an offline conversation with @sdaftuar here (with regard to #16847) which I think is relevant. Dynamically flipping `nLocalServices` bits doesn't make a ton of sense because we'd need to disconnect and reconnect to all of our peers in order for the flip to have any effect.

When thinking about this, it's instructive to look at how IBD works: we don't flip `NODE_NETWORK` off and on once we've started and finished IBD; instead net_processing just behaves differently during IBD and peers are able to gracefully handle nodes that signal NODE_NETWORK but haven't completed the sync yet and therefore drop getheaders messages.

I think the way forward in this case is to do something analogous: signal `NODE_COMPACT_FILTERS` from startup when appropriate and ensure that nodes seeking block filters account for the fact that a peer they request filters from may be in the process of building them.

Reply:
> Dynamically flipping nLocalServices bits doesn't make a ton of sense because we'd need to disconnect and reconnect to all of our peers in order for the flip to have any effect.

I disagree with this. New inbound connections will see that the service bit has been flipped. And clients who need the filters wouldn't connect until it is on, so it's not like any inbound connections need to periodically disconnect/reconnect. And if I understand correctly, we re-advertise our local address with service bits to each peer every day on average, so the network should see the change eventually without the node operator having to do anything.

>  instead net_processing just behaves differently during IBD and peers are able to gracefully handle nodes that signal NODE_NETWORK but haven't completed the sync yet and therefore drop getheaders messages.

This feels a bit different because the `version` message includes a start height indicating how far behind the node is. I suppose clients could also just request checkpoints and determine how far behind a node's filter index is at least to within 1,000 blocks though.

Reply:
In any case, can you simply modify `nLocalServices` directly when the index has finished syncing? Or does other code that accesses it directly need to see the bit set even before sync is done?

Reply:
I have removed the dynamic flipping in accordance with reviewers' suggestions.